### DISCUSSION TITLE
Getting a formula written to build haskell-language-server

### DISCUSSION THREAD

**Discussion by ELLIOTTCABLE** [Homebrew/discussions#386] [Tap maintenance and Homebrew development]
I'm attempting to write a formula for [`haskell-language-server`](https://github.com/haskell/haskell-language-server); I haven't written a formula in many years, so please forgive me if I say anything dumb!

HLS requires *several* downloads, per version: https://github.com/haskell/haskell-language-server/releases/tag/0.7.1

- the `haskell-language-server-wrapper` binary,
- and an `haskell-language-server-8.8.2` (or similar) binary for each version of GHC being used in conjunction.

Realistically, I think the best solution is to download *all* of them — there's only a half-dozen or so, and I don't think having a few `haskell-language-server-macOS-N.N.N` copies in `/usr/local/bin` is a huge deal.

That said, is there a best way to go about this in a formula? Any directions or advice? I don't want to do a huge hack if I don't have to!

**Comment by carlocab** [comment]
I _think_ the answer here is that you're not meant to download any of them: https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae

The only exception is when the binary you're downloading is for the purposes of bootstrapping at build-time. These binaries are thrown away when the build is complete. (Still, we try to avoid these where we can.) If that's the case here, I don't know if there's a better way than using multiple `resource` blocks.

**Reply by ELLIOTTCABLE** [reply]
Hm, that's a fair point.

Yuck, that's really unfortunate; `haskell-language-server`'s build-system is pretty complicated, and not-at-all friendly to something like Homebrew.

Thanks for the reply!

**Reply by carlocab** [reply]
Sure. You may wish to contribute a `haskell-language-server` cask instead. Binary-only casks are fine. This is how `julia` is currently offered on Homebrew, since the Julia build system is also very complicated and not Homebrew-friendly.

**Reply by kret** [reply]
How about a cask roughly like this: https://github.com/kret/homebrew-cask/commit/e14dd52c99010935f1c4adebe2c9f07eb004f57a?
...just download and decompress a full macOS release, with wrapper and all versions?

I did not specify there the dependencies as it seems to me that there are several, different, preferred setups of GHC/stack/cabal/etc that specifying dependencies might actually break some devs' environment. Could be updated, of course.
At the moment `ghc` formula is at version `8.10.3` and `hls 0.7.1` does not provide a version for it (https://github.com/haskell/haskell-language-server/releases/tag/0.7.1). `hls 0.8.0` should provide this support, and be released soon, though.

This cask seems to work with nvim, command line execution.
However, in order to persuade VSCode to use this `hls` I had to install the cask with `brew install haskell-language-server --no-quarantine` or suffer the manual macOS security overrides in `System Preferences > Security & Privacy`.
I did not find a ready solution for putting the `--no-quarantine` into the cask itself.
I could use `caveats` to inform users of `--no-quarantine`? Seems to be a bit too late, though.
Not sure how to best approach this.

Please let me know what you think, or just feel free to use/expand this solution :)

**Reply by kret** [reply]
Updated cask on my branch to `hls 0.8.0` https://github.com/kret/homebrew-cask/commit/7d7acb3699bbfc29ea399b8d7c1ee5f8c4e8f7d2. Now it covers GHC 8.10.3 so same version as `ghc` Homebrew formula.
Should be easier now to add `depends_on`, if you preferred that.

**Reply by kret** [reply]
...more humane link to proposed cask file: https://github.com/kret/homebrew-cask/blob/haskell-language-server-cask/Casks/haskell-language-server.rb

**Reply by carlocab** [reply]
Note that this may not be an acceptable Cask: https://github.com/homebrew/homebrew-cask/blob/master/doc/faq/rejected_casks.md

>The app is both open-source and CLI-only (i.e. it only uses the `binary` artifact). In that case, and [in the spirit of deduplication](https://github.com/Homebrew/homebrew-cask/issues/15603), submit it first to [Homebrew/core](https://github.com/Homebrew/homebrew-core) as a formula that builds from source. If it is rejected, you may then try again as a cask (link us to the issue so we can see the discussion and reasoning for rejection).

**Reply by kret** [reply]
So far I have succeeded with...
- a *cask* for a full `hls 0.8.0` release: https://github.com/kret/homebrew-hls/blob/main/Casks/haskell-language-server-all.rb,
- a *formula* for `hls 0.8.0` with GHC provided by homebrew (8.10.3) https://github.com/kret/homebrew-hls/blob/main/Formula/haskell-language-server.rb.

**Cask**
- :+1: all-in-one package, as built by `hls` maintainers,
- :-1: needs to be installed with `--no-quarantine` switch or user needs to manually
  confirm security exception for each binary in System Preferences
  as they are being executed first time :(

Is there a way to make this cask not-`--no-quarantine`able?

**Formula**
- :+1: follows `Homebrew` guidelines,
- :-1: uses a build command ripped out of https://github.com/haskell/haskell-language-server/blob/3464ed8b6ba1fbc443dae95f9edb2e07420b1b26/install/src/Cabal.hs#L61
  which is very fragile and would need to be updated every time `hls` maintainers update their build setup,
- :-1: is probably not useful to Haskellers that work on any project with GHC version different from `Homebrew`-provided one

Feel free to comment.

**Comment by ELLIOTTCABLE** [comment]
Okay, I've done a little more investigation into this. I'm a newbie, so I don't know if I'll be able to complete this process myself, but maybe somebody coming through later can benefit from my investigation.

The problem here is mostly that:

1. `haskell-language-server` needs to be built for the *specific* GHC version it's going to be used with.
2. End-users may have any GHC revision installed, and expect `haskell-language-server` to work with that revision.

Upstream, this is solved by shipping `haskell-language-server-wrapper`, along with a bunch of pre-built `haskell-language-server-N.N.N` binaries that work with a specific GHC version. (As described in my misguided OP, heh.)

Further complicating this, is "what GHC versions is a given Joe Enduser on Homebrew using" is actually non-trivial:

1. They may have `brew install ghc`'d, and thus have a bare, specific version of GHC installed; but *not* have Cabal or Stack installed
2. They may have `brew install stack`'d, and their Stack install may have one-or-more versions of GHC installed thru Stack,
3. Or they may have installed `stack` or `ghcup` or other tooling, and not used Homebrew *at all* to install Haskell, while still expecting `brew install haskell-language-server` to work without intervention.

Upstream, the project handles this with a [custom build-system that queries Stack and cabal](https://github.com/haskell/haskell-language-server/blob/master/install/src/HlsInstall.hs), attempting to determine what versions of GHC it should build against. /=

So, tl;dr I'm not sure how to solve this in a single Formula, that we can put in homebrew-core, that supports all three of the above categories. I'm guessing we'd have to have CI build *all* of the possible GHC versions, so that the single `haskell-language-server` bottle contains whichever specific binar(y|ies) a given user may need ... but I don't know the ~blessed way~ to write such a formula. (In particular, the naive way I can think of doing that, would result in a user who *doesn't* use bottles, building a dozen versions of GHC, from source, in serial, to build a dozen versions of `haskell-language-server`, that they don't need ... yikes. Especially-yikes given how famously slow GHC is to build.)

I hope somebody smarter than me, or more familiar with Haskell than me, or more familiar with Homebrew than me, can figure this out!

----

Here's my initial naive attempt, although I haven't had much luck getting it to work:

```ruby
class HaskellLanguageServer < Formula
  desc "Integration point for ghcide and haskell-ide-engine. One IDE to rule them all."
  homepage "https://github.com/haskell/haskell-language-server"
  license "Apache-2.0"

  # Use the tag instead of the tarball to get submodules
  url "https://github.com/haskell/haskell-language-server.git",
      tag:      "0.7.1",
      revision: "e4f677e1780fe85a02b99a09404a0a3c3ab5ce7c"

  head "https://github.com/haskell/haskell-language-server.git"

  livecheck do
    url :stable
    strategy :github_latest
  end

  depends_on "cabal-install" => :build
  depends_on "ghc" => :build
  depends_on "icu4c"

  def install
    system "cabal", "v2-update"
    system "cabal", "v2-run", *std_cabal_v2_args,
      "./install.hs", "--project-file", "install/shake.project"
  end
end
```

**Reply by gromgit** [reply]
> I'm not sure how to solve this in a single Formula, that we can put in homebrew-core, that supports all three of the above categories.

I wouldn't worry about that. Since you say HLS is tightly coupled with GHC, focus on getting your formula working with the current `ghc` formula (I assume that the latest release in both projects at any time are guaranteed to be compatible). Trying to do what you describe is like leaping from "weekend jogger" to "Iron Man pack-leader" in a day -- you'll just drown.

**Reply by carlocab** [reply]
Ok, let's just try to simplify this problem a bit. Guaranteeing compatibility with software that wasn't installed using Homebrew is difficult and untestable (in particular, not tested by Homebrew CI), so let's consider that nice-to-have-but-not-necessary.

>They may have brew install ghc'd, and thus have a bare, specific version of GHC installed; but not have Cabal or Stack installed

We can add hard dependencies to `cabal` and `stack` if they're necessary. I think the goal here at minimum, would be to get the language server to work if the user has Homebrew `ghc` installed. If it'll work with Homebrew `ghc@8.8` and `ghc@8.6` too, even better.

**Comment by jneira** [comment]
haskell-language-server is on hackage now and can be installed in a more friendly way with cabal install: https://hackage.haskell.org/package/haskell-language-server

**Comment by kret** [comment]
Solution suggestion: https://github.com/Homebrew/homebrew-cask/pull/100527.
This downloads a full official HLS macOS release and makes available on path.

### Inconveniences
* macOS security warnings will popup once each binary is run. This can be worked around using Homebrew's `--no-quarantine` flag while re/-installing the cask.
* There are no dependencies on Homebrew's GHC/Cabal/Stack specified because it is my understanding that there are several way for projects to fetch those.

**Alternative solution** might be to create a cask for each binary (available as part official releases too: https://github.com/haskell/haskell-language-server/releases/tag/1.0.0) so that users can pick specific ones...

**Reply by jneira** [reply]
mmm not sure how much time cost download all of them but it is the simplest solution so i would go for it, you can always create independent cask for specific versions afterwards (maybe not all of them but the more popular ones?)

**Reply by kret** [reply]
Well, PR was rejected, as suggested earlier on this discussion 😄 (https://github.com/Homebrew/homebrew-cask/pull/100527#issuecomment-789941713).
As to the time to download and unpack the full HLS cask, it was about 20s on my laptop (not the fastest internet connection). I would not mind, 20s and I could use HLS in any project supported by it 🚀  🪐
